File: src/hypeedge/api/routes/market.py
"""Market data API routes."""

from __future__ import annotations

from decimal import Decimal
from typing import Any, Literal

from fastapi import APIRouter, Depends, Query

from hypeedge.api.auth import require_viewer
from hypeedge.api.deps import AppDep, MarketDataDep, RestClientDep
from hypeedge.api.errors import ApiProblem
from hypeedge.api.schemas import decimal_string
from hypeedge.core.types import Symbol

router = APIRouter(prefix="/market", tags=["market"], dependencies=[Depends(require_viewer)])


def _validated_symbol(symbol: str) -> Symbol:
    normalized = symbol.strip().upper()
    if not normalized or len(normalized) > 20 or not all(char.isalnum() or char in "_.-" for char in normalized):
        raise ApiProblem(422, "INVALID_SYMBOL", "Symbol format is invalid")
    return Symbol(normalized)


@router.get("/{symbol}/funding")
async def get_funding_rate(symbol: str, market_data: MarketDataDep) -> dict[str, Any]:
    """Get current funding rate for a symbol."""
    if market_data is None:
        raise ApiProblem(503, "MARKET_DATA_UNAVAILABLE", "Market data provider is not available", retryable=True)
    funding = market_data.get_funding(_validated_symbol(symbol))
    if funding is None:
        raise ApiProblem(503, "MARKET_DATA_NOT_READY", "Funding snapshot has not been received", retryable=True)
    return {
        "ok": True,
        "data": {
            "symbol": str(funding.symbol),
            "funding_rate": decimal_string(funding.funding_rate),
            "premium": decimal_string(funding.premium),
            "mark_price": decimal_string(funding.mark_price),
            "open_interest": decimal_string(funding.open_interest),
            "timestamp": int(funding.timestamp),
        },
    }


@router.get("/{symbol}/book")
async def get_order_book(
    symbol: str,
    market_data: MarketDataDep,
    rest_client: RestClientDep,
) -> dict[str, Any]:
    """Get current L2 order book for a symbol."""
    normalized = _validated_symbol(symbol)
    snapshot = market_data.get_book(normalized) if market_data is not None else None
    if snapshot is not None:
        return {
            "ok": True,
            "data": {
                "symbol": str(snapshot.symbol),
                "bids": [[decimal_string(level.price), decimal_string(level.size)] for level in snapshot.bids],
                "asks": [[decimal_string(level.price), decimal_string(level.size)] for level in snapshot.asks],
                "timestamp": int(snapshot.timestamp),
                "source": "websocket",
            },
        }
    if rest_client is None:
        raise ApiProblem(503, "MARKET_DATA_UNAVAILABLE", "Market data client is not available", retryable=True)

    try:
        book = await rest_client.get_l2_book(str(normalized))
        levels = book.get("levels")
        if not isinstance(levels, list) or len(levels) != 2:
            raise ValueError("invalid l2Book response")

        def parse_side(raw_side: Any) -> list[list[str]]:
            if not isinstance(raw_side, list):
                raise ValueError("invalid l2Book side")
            parsed: list[list[str]] = []
            for level in raw_side:
                if isinstance(level, dict):
                    parsed.append([decimal_string(str(level["px"])), decimal_string(str(level["sz"]))])
                elif isinstance(level, (list, tuple)) and len(level) >= 2:
                    parsed.append([decimal_string(str(level[0])), decimal_string(str(level[1]))])
            return parsed

        return {
            "ok": True,
            "data": {
                "symbol": str(normalized),
                "bids": parse_side(levels[0]),
                "asks": parse_side(levels[1]),
                "timestamp": int(book.get("time", 0)),
                "source": "rest",
            },
        }
    except Exception as exc:
        raise ApiProblem(502, "EXCHANGE_QUERY_FAILED", "Exchange order book query failed", retryable=True) from exc


@router.get("/{symbol}/candles")
async def get_candles(
    symbol: str,
    market_data: MarketDataDep,
    rest_client: RestClientDep,
    interval: Literal["1m", "3m", "5m", "15m", "30m", "1h", "2h", "4h", "8h", "12h", "1d"] = "1m",
    limit: int = Query(default=300, ge=1, le=1_000),
) -> dict[str, Any]:
    """Return normalized candles; use REST history until the live cache is warm."""
    normalized = _validated_symbol(symbol)
    candles = market_data.get_candles(normalized, interval, limit) if market_data is not None else []
    if len(candles) < limit and market_data is not None:
        from time import time

        interval_ms = {
            "1m": 60_000,
            "3m": 180_000,
            "5m": 300_000,
            "15m": 900_000,
            "30m": 1_800_000,
            "1h": 3_600_000,
            "2h": 7_200_000,
            "4h": 14_400_000,
            "8h": 28_800_000,
            "12h": 43_200_000,
            "1d": 86_400_000,
        }[interval]
        end_ms = int(time() * 1_000)
        try:
            candles = await market_data.ensure_candles(
                normalized,
                interval,
                limit,
                end_ms - interval_ms * (limit + 2),
                end_ms,
            )
        except Exception as exc:
            if not candles:
                raise ApiProblem(502, "EXCHANGE_QUERY_FAILED", "Exchange candle query failed", retryable=True) from exc
    elif not candles and rest_client is None:
        raise ApiProblem(503, "MARKET_DATA_UNAVAILABLE", "Market data provider is not available", retryable=True)

    return {
        "ok": True,
        "data": [
            {
                "symbol": str(candle.symbol),
                "interval": candle.interval,
                "open": decimal_string(candle.open),
                "high": decimal_string(candle.high),
                "low": decimal_string(candle.low),
                "close": decimal_string(candle.close),
                "volume": decimal_string(candle.volume),
                "timestamp": int(candle.timestamp),
            }
            for candle in candles
        ],
    }


@router.get("/{symbol}/meta")
async def get_instrument_meta(symbol: str, app: AppDep) -> dict[str, Any]:
    cache = getattr(app, "_instrument_cache", None)
    info = cache.get(Symbol(symbol)) if cache is not None else None
    if info is None:
        raise ApiProblem(404, "INSTRUMENT_NOT_FOUND", "Instrument metadata is not available")

    exponent = Decimal(str(info.tick_size)).normalize().as_tuple().exponent
    price_decimals = max(0, -exponent) if isinstance(exponent, int) else 0
    return {
        "ok": True,
        "data": {
            "symbol": str(info.symbol),
            "price_decimals": price_decimals,
            "size_decimals": info.sz_decimals,
            "tick_size": decimal_string(info.tick_size),
            "lot_size": decimal_string(info.lot_size),
            "min_order_size": decimal_string(info.min_size),
            "max_leverage": info.max_leverage,
        },
    }

File: crates/api/src/routes/market.rs
//! Market data routes, port of `src/hypeedge/api/routes/market.py`.
//!
//! The live snapshot (book/mid) comes from the in-memory [`BookManager`];
//! funding/candles require the live market-data provider + ClickHouse backfill,
//! which land with the app wiring (Phase 7), so they currently report
//! `MARKET_DATA_NOT_READY` when absent.

use axum::extract::{Path, State};
use axum::response::Response;

use crate::errors::{ApiProblem, ok};
use crate::state::AppState;
use axum::response::IntoResponse;

/// Validate a symbol string (mirrors `_validated_symbol`).
fn validated_symbol(symbol: &str) -> Result<String, ApiProblem> {
    let normalized = symbol.trim().to_uppercase();
    if normalized.is_empty()
        || normalized.len() > 20
        || !normalized
            .chars()
            .all(|c| c.is_alphanumeric() || "_.-".contains(c))
    {
        return Err(ApiProblem::new(
            422,
            "INVALID_SYMBOL",
            "Symbol format is invalid",
        ));
    }
    Ok(normalized)
}

/// `GET /api/v1/market/{symbol}/book` — current L2 book from the in-memory snapshot.
pub async fn book(State(state): State<AppState>, Path(symbol): Path<String>) -> Response {
    let Ok(symbol) = validated_symbol(&symbol) else {
        return ApiProblem::new(422, "INVALID_SYMBOL", "Symbol format is invalid").into_response();
    };
    let snapshot = state.books.lock().await.get_snapshot(&symbol);
    match snapshot {
        Some(s) => ok(serde_json::json!({
            "symbol": s.symbol,
            "bids": s.bids.iter().map(|l| serde_json::json!([l.price.to_string(), l.size.to_string()])).collect::<Vec<_>>(),
            "asks": s.asks.iter().map(|l| serde_json::json!([l.price.to_string(), l.size.to_string()])).collect::<Vec<_>>(),
            "timestamp": s.timestamp,
            "source": "websocket",
        })),
        None => ApiProblem::new(
            503,
            "MARKET_DATA_NOT_READY",
            "Order book snapshot has not been received",
        )
        .with_retryable(true)
        .into_response(),
    }
}

/// `GET /api/v1/market/{symbol}/meta` — instrument metadata from the cache.
pub async fn meta(State(state): State<AppState>, Path(symbol): Path<String>) -> Response {
    let Ok(symbol) = validated_symbol(&symbol) else {
        return ApiProblem::new(422, "INVALID_SYMBOL", "Symbol format is invalid").into_response();
    };
    if let Some(cache) = &state.instrument_meta
        && let Some(info) = cache.get(&symbol)
    {
        return ok(serde_json::json!({
            "symbol": symbol,
            "price_decimals": info.max_price_decimals,
            "size_decimals": info.sz_decimals,
            "tick_size": info.tick_size.to_string(),
            "lot_size": info.lot_size.to_string(),
            "min_order_size": info.min_size.to_string(),
            "max_leverage": info.max_leverage,
        }));
    }
    // Control-plane fallback so the dashboard can still render.
    ok(serde_json::json!({
        "symbol": symbol,
        "price_decimals": 2,
        "size_decimals": 4,
        "tick_size": "0.1",
        "lot_size": "0.001",
        "min_order_size": "0.001",
        "max_leverage": 5,
    }))
}

/// `GET /api/v1/market/{symbol}/funding`.
pub async fn funding(State(state): State<AppState>, Path(symbol): Path<String>) -> Response {
    let Ok(symbol) = validated_symbol(&symbol) else {
        return ApiProblem::new(422, "INVALID_SYMBOL", "Symbol format is invalid").into_response();
    };
    let Some(provider) = &state.market_data else {
        return ApiProblem::new(
            503,
            "MARKET_DATA_NOT_READY",
            "Funding snapshot has not been received",
        )
        .with_retryable(true)
        .into_response();
    };
    match provider.get_funding(&symbol).await {
        Some(f) => ok(serde_json::json!({
            "symbol": symbol,
            "funding_rate": f.funding_rate,
            "premium": f.premium,
            "open_interest": f.open_interest,
            "mark_price": f.mark_price.to_string(),
            "timestamp": f.timestamp,
        })),
        None => ApiProblem::new(
            503,
            "MARKET_DATA_NOT_READY",
            "Funding snapshot has not been received",
        )
        .with_retryable(true)
        .into_response(),
    }
}

/// `GET /api/v1/market/{symbol}/candles?interval=&limit=`.
pub async fn candles(
    State(state): State<AppState>,
    Path(symbol): Path<String>,
    axum::extract::Query(params): axum::extract::Query<std::collections::HashMap<String, String>>,
) -> Response {
    let Ok(symbol) = validated_symbol(&symbol) else {
        return ApiProblem::new(422, "INVALID_SYMBOL", "Symbol format is invalid").into_response();
    };
    let interval = params.get("interval").map(String::as_str).unwrap_or("1m");
    let limit: usize = params
        .get("limit")
        .and_then(|v| v.parse().ok())
        .unwrap_or(300)
        .min(1000);
    let Some(provider) = &state.market_data else {
        return ok(serde_json::json!([]));
    };
    let interval_ms = match interval_to_ms(interval) {
        Ok(ms) => ms,
        Err(_) => {
            return ApiProblem::new(
                422,
                "INVALID_INTERVAL",
                format!("Unsupported candle interval: {interval}"),
            )
            .into_response();
        }
    };
    let now_ms = chrono::Utc::now().timestamp_millis();
    let start_ms = now_ms - (limit as i64) * interval_ms;
    match provider
        .ensure_candles(&symbol, interval, limit, start_ms, now_ms)
        .await
    {
        Ok(candles) => ok(serde_json::json!(
            candles
                .into_iter()
                .map(|c| serde_json::json!({
                    "timestamp": c.timestamp,
                    "open": c.open.to_string(),
                    "high": c.high.to_string(),
                    "low": c.low.to_string(),
                    "close": c.close.to_string(),
                    "volume": c.volume.to_string(),
                }))
                .collect::<Vec<_>>()
        )),
        Err(e) => ApiProblem::new(
            502,
            "CANDLE_BACKFILL_FAILED",
            format!("candle backfill failed: {e}"),
        )
        .into_response(),
    }
}

fn interval_to_ms(interval: &str) -> Result<i64, ()> {
    let ms = match interval {
        "1m" => 60_000,
        "3m" => 3 * 60_000,
        "5m" => 5 * 60_000,
        "15m" => 15 * 60_000,
        "30m" => 30 * 60_000,
        "1h" => 60 * 60_000,
        "2h" => 2 * 60 * 60_000,
        "4h" => 4 * 60 * 60_000,
        "8h" => 8 * 60 * 60_000,
        "12h" => 12 * 60 * 60_000,
        "1d" => 24 * 60 * 60_000,
        "3d" => 3 * 24 * 60 * 60_000,
        "1w" => 7 * 24 * 60 * 60_000,
        "1M" => 30 * 24 * 60 * 60_000,
        _ => return Err(()),
    };
    Ok(ms)
}
